src/libcore/iter.rs

```diff
@@ -1565,24 +1565,13 @@ pub struct Map<I, F> {
     f: F,
 }
 
-impl<I: Iterator, F, B> Map<I, F> where F: FnMut(I::Item) -> B {
-    #[inline]
-    fn do_map(&mut self, elt: Option<I::Item>) -> Option<B> {
-        match elt {
-            Some(a) => Some((self.f)(a)),
-            _ => None
-        }
-    }
-}
-
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<B, I: Iterator, F> Iterator for Map<I, F> where F: FnMut(I::Item) -> B {
     type Item = B;
 
     #[inline]
     fn next(&mut self) -> Option<B> {
-        let next = self.iter.next();
-        self.do_map(next)
+        self.iter.next().map(|a| (self.f)(a))
```

Comment:
I believe this should be possible with `self.iter.next().map(&mut self.f)`. The `fn` types are distinct from the `Fn` traits because `fn` is a specific type for a function pointer (but implements the `Fn` trait).

We cannot implement `FnMut` for `&FnMut`, but it is implemented for `&mut FnMut` (as the mutable borrow is required).